ddtrace/debugging/_origin/span.py

```diff
@@ -0,0 +1,248 @@
+    def __init__(self, f):
+        super().__init__(f)
+
+        lines = linenos(f)
+        start_line = min(lines)
+        filename = str(Path(f.__code__.co_filename).resolve())
+        name = f.__qualname__
+        module = f.__module__
+        self.location = EntrySpanLocation(
+            name=name,
+            start_line=start_line,
+            end_line=max(lines),
+            file=filename,
+            module=module,
+            probe=t.cast(EntrySpanProbe, EntrySpanProbe.build(name=name, module=module, function=name)),
+        )
+
+    def __enter__(self):
+        super().__enter__()
+
+        root = ddtrace.tracer.current_root_span()
+        location = self.location
+        if root is None or root.get_tag("_dd.entry_location.file") is not None:
+            return
+
+        # Add tags to the local root
+        root.set_tag_str("_dd.entry_location.file", location.file)
+        root.set_tag_str("_dd.entry_location.start_line", str(location.start_line))
+        root.set_tag_str("_dd.entry_location.end_line", str(location.end_line))
+        root.set_tag_str("_dd.entry_location.type", location.module)
+        root.set_tag_str("_dd.entry_location.method", location.name)
+
+        if config._trace_span_origin_enriched:
+            # Create a snapshot
+            snapshot = Snapshot(
+                probe=location.probe,
+                frame=self.__frame__,
+                thread=current_thread(),
+                trace_context=root,
+            )
+
+            # Capture on entry
+            context = Debugger.get_collector().attach(snapshot)
+
+            # Correlate the snapshot with the span
+            root.set_tag_str("_dd.entry_location.snapshot_id", snapshot.uuid)
+
+            self.set("context", context)
+            self.set("start_time", compat.monotonic_ns())
+
+    def _close_context(self, retval=None, exc_info=(None, None, None)):
+        try:
+            context: SignalContext = self.get("context")
+        except KeyError:
+            # No snapshot was created
+            return
+
+        context.exit(retval, exc_info, compat.monotonic_ns() - self.get("start_time"))
+
+    def __return__(self, retval):
+        self._close_context(retval=retval)
+        return super().__return__(retval)
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self._close_context(exc_info=(exc_type, exc_value, traceback))
+        super().__exit__(exc_type, exc_value, traceback)
+
+
+@attr.s
+class SpanOriginProcessor(SpanProcessor):
+    _instance: t.Optional["SpanOriginProcessor"] = None
+
+    def on_span_start(self, span: Span) -> None:
+        if span.span_type not in EXIT_SPAN_TYPES:
+            return
+
+        # Add call stack information to the exit span. Report only the part of
+        # the stack that belongs to user code.
+        # TODO: Add a limit to the number of frames to capture
+        seq = count(1)
+        for frame in frame_stack(sys._getframe(1)):
```

Comment:
did we decided to add the entire stack for exit spans?
if so, then how do we deal the frame numbering? currently is running numbers of user-code frames. but if we also report the entire stack, maybe those should match the frame depth?